### ISSUE TITLE
Modifying simulation temperature, pressure, and parameters in an existing Context

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#1440]
I'm trying to make sure I have the correct scheme for modifying temperature, pressure, and global parameters within an existing simulation `Context` and `Integrator`, and would appreciate suggestions/advice.  There are a number of complications due to interface nonuniformity, and I'm not even sure if there's really a way to do this correctly at all.

When we want to modify the **temperature**, we have to ensure that we update
* OpenMM integrators, which unfortunately do not all have a `setTemperature()` API call; 
* If we are using a `CustomIntegrator`s, which would have to define/expose the temperature through a global parameter, but I don't think these global parameters can actually be updated via a call to `Context.setParameter()` since this only affects `Force` parameters (?)
* `MonteCarloBarostat`, which has a `setTemperature()` method, but I am unsure of whether this updates the temperature used within the `Context`; unfortunately, the API is also different from setting the pressure, which is done via a global context parameter whose name is retrieved via `MonteCarloBarostat.Pressure()`
* if any `Force` objects used a `temperature` global parameter (such as a temperature-dependent GB model), we'd have to update the `temperature` global context parameter via `Context.setParameter()`

When we want to modify the **pressure**, we would update
* `MonteCarloBarostat`, which exposes the pressure via a global context parameter whose name is retrieved via `MonteCarloBarostat.Pressure()`

When we want to update simulation context parameters that affect the energy computation, this is actually very straightforward: We can call `Context.setParameter().

Here's my best attempt at putting all of this together into code:
```
        if integrator is not None:
            if hasattr(integrator, 'setTemperature'):
                integrator.setTemperature(self.temperature)
            if hasattr(integrator, 'getNumGlobalVariables'):
                integrator_parameters = [ integrator.getGlobalVariableName(index) for index in range(integrator.getNumGlobalVariables()) ]
                if 'temperature' in integrator_parameters:
                    integrator.setGlobalVariableByName('temperature', self.temperature.value_in_unit_system(unit.md_unit_system))
                if 'kT' in integrator_parameters:
                    integrator.setGlobalVariableByName('kT', self.kT.value_in_unit_system(unit.md_unit_system))

        # Set pressure
        if self.pressure:
            forces = { self.system.getForce(index).__class__.__name__ : self.system.getForce(index) for index in range(self.system.getNumForces()) }
            barostat = forces['MonteCarloBarostat']
            barostat.setTemperature(self.temperature) # DOES THIS ACTUALLY UPDATE THE TEMPERATURE?
            context.setParameter(barostat.Pressure(), self.pressure.value_in_unit_system(unit.md_unit_system))

        # Set Context parameters
        if self.parameters is not None:
            for parameter in self.parameters:
                value = self.parameters[parameter]
                if hasattr(value, 'unit'):
                    value = value.value_in_unit_system(unit.md_unit_system)
                context.setParameter(parameter, value)
```
As noted, I'm not entirely sure whether the call to `barotstat.setTemperature()` actually updates the temperature used in the simulation.

Beyond just ensuring this will correctly update the simulation temperature, pressure, and parameters, it would also be great to think about how we might be able to simplify this process by harmonizing interfaces. Some concrete ideas:
* Extend the `Integrator` API to include `setTemperature()` and `getTemperature()` that would store the temperature internally. In `CustomIntegrator`s, this could expose a special global variable `temperature` that could be accessed to get the current temperature. 
* Have `MonteCarloBarostat` also have a `temperature` context parameter that is accessible the same way that `barostat.Pressure()` is, or alternatively, have a `barostat.setPressure()` that updates the pressure in the `Context` if one is associated.
* I'm not sure if the Python wrappers to `context.setParameter()` automatically strip units in the correct way.

**Comment by Lnaden** [comment]
`CustomIntegrator`s do have `setGlobalVariable` API call if you have the temperature defined through a integrator-only variable. Otherwise `CustomIntegrator`s do have in access to the `Context` variables namespace, so if you have a temperature defined there, the `CustomIntegrator` can just use that variable by name in place.

> I'm not sure if the Python wrappers to context.`setParameter()` automatically strip units in the correct way.

I know `setParameter` does remove units, but are you referring to stripping the units to the correct internal representation? E.g. you give it a temp in °C and it converts it to K (unitless) internally? I'm not sure if it does that, but I imagine it would be easy to test.

**Comment by jchodera** [comment]
> CustomIntegrators do have setGlobalVariable API call if you have the temperature defined through a integrator-only variable.

I use that capability in my example above, but `temperature` is so common that it would be very useful to have this able to be set/get through a uniform API at the `Integrator` level.

> Otherwise CustomIntegrators do have in access to the Context variables namespace, so if you have a temperature defined there, the CustomIntegrator can just use that variable by name in place.

This could work too, but seems to make things even more confusing.

> I know setParameter does remove units, but are you referring to stripping the units to the correct internal representation? E.g. you give it a temp in °C and it converts it to K (unitless) internally? I'm not sure if it does that, but I imagine it would be easy to test.

Right---I was wondering if it converted to `md_unit_system` internally.  You're right that this is easy to check.

**Comment by peastman** [comment]
`MonteCarloBarostat` is definitely inconsistent with everything else.  It *should* have a `setDefaultTemperature()` method and then define a context parameter, just like it does for pressure.  Instead, it just has `setTemperature()` which immediately affects all Contexts created from that system.  *Not* how Forces are supposed to work!

I'd definitely like to change that.  It would necessarily break backward compatibility though.  Hopefully this is a rare case so that wouldn't affect many people.  What do you think?

**Comment by Lnaden** [comment]
+1 for unifying the force objects. 

Would it be worth having all the objects which need a temperature hook into a common Context parameter so that there is only one variable that needs to be updated? The trade off though is it becomes harder to create systems where some degrees of freedom are raised to different temperatures than others for accelerated sampling. Not very common, but it has been done.

Since this would be a backwards compatibility break, I think a change should be relegated to 7.1 or 7.2 if we go forward with it.

**Comment by jchodera** [comment]
I am still not actually sure there was a correct way to update the temperature and pressure in the first place, so I am totally ok with breaking backwards compatibility to sort out a straightforward way to do this.

**Comment by Lnaden** [comment]
@jchodera Can confirm passing in any `Quantity` from the `unit` module converts it to the `md_unit_system` internally. However, there does not appear to be the units Celsius, Fahrenheit or any other temperature unit in the `unit` module, so we don't have to worry about passing that in.

**Comment by peastman** [comment]
It does have temperature units, but only ones that put 0 at absolute 0.  So there are two of them: Kelvin and Planck units.

**Comment by swails** [comment]
It's missing the [Kelvin-equivalent of the Fahrenheit system](https://en.wikipedia.org/wiki/Rankine_scale) :).

It would be a real challenge to support "arbitrary" temperature units.  Unlike *every* other unit conversion, which can be done via a simple multiplicative scaling, conversions between Celcius and Kelvin (or Fahrenheit and Celcius or Kelvin) relies on a constant shift.  As a result, conversions are not always well-defined: 10 deg. C *can* convert to either 282 K *or* 10 K, depending on whether it was a difference or a absolute temperature.  That's another way of saying that you lose the feature of associativity:

```Python
(100 - 90) * Celcius # Would evaluate to 282 K ?
100*Celcius - 90*Celcius # Would evaluate to 10 K ?
```

Also, composite units including any non-absolute temperature unit are nonsensical.  Simple example: what is the equivalent of Boltzmann's constant in J/mol*deg.C? :)

**Comment by jchodera** [comment]
Bumping this issue to see if we might be able to standardize the correct way to update temperature and pressure for an existing `Context`.  I'm still not sure I have this correct...

**Comment by jchodera** [comment]
Following #1468, I'm continuing to brainstorm about how we can make it easier for users to correctly set the temperature and pressure without allowing the potential for making simple mistakes that lead to incorrect simulations.

One thing that I think would greatly simplify things is to unify the way we set the temperature in both standard and custom `Integrator`s.  If we extend the API for [`Integrator`](https://github.com/pandegroup/openmm/blob/master/openmmapi/include/openmm/Integrator.h) by adding `getTemperature()` and `setTemperature()` at the base `Integrator` level:
```C++
    /**
     * Get the temperature of the heat bath (in Kelvin).
     *
     * @return the temperature of the heat bath, measured in Kelvin
     */
    double getTemperature() const {
        return temperature;
    }
    /**
     * Set the temperature of the heat bath (in Kelvin).
     * If the Integrator is associated with a Context, the new temperature will be used the next time `step()` is called.
     * Note that not all integrators may use the temperature variable.
     *
     * @param temp    the temperature of the heat bath, measured in Kelvin
     */
    void setTemperature(double temp) {
        temperature = temp;
    }
```
and making `temperature` a private variable.  This would allow us to eliminate the need for `MonteCarloBarostat` to have its own temperature variable or be initialized with a temperature, since it can retrieve the current temperature from the `Context` with which the integrator is associated whenever the Monte Carlo moves occur.  The one drawback would be that the API for `MonteCarloBarostat` currently includes the temperature, but this argument could be ignored and a new constructor that omits this argument introduced, with the old constructor eventually deprecated. 

For `CustomIntegrator` objects, we could expose `temperature` and `kT = kB * temperature` as a new variable (both in `md_unit_system` units). The `CustomIntegrator` API currently only includes timestep, but we could introduce a two-parameter constructor that also allows the temperature to be specified if we thought this was important---but it is probably unnecessary since `integrator.setTemperature(temp)` can be called instead.

This would then leave a unique recommended way of updating temperature---via `integrator.setTemperature(temp)`---that will ensure both the barostat uses the correct temperature and `CustomIntegrators` can have their temperatures updated on the fly (which was left up to users to specify as global context parameters previously).  The pressure still must be updated via a context parameter `pressure`, but perhaps the addition of a `MonteCarloBarostat.updatePressureInContext(context)` convenience method could also help there.

**Comment by peastman** [comment]
> This would then leave a unique recommended way of updating temperature

Except it still wouldn't be unique.  AndersenThermostat would still be different.  And then what about DrudeLangevinIntegrator, which has two different temperatures?  I'm also kind of uncomfortable about putting a `setTemperature()` method onto integrators that don't have anything to do with temperature, like VerletIntegrator.

**Comment by jchodera** [comment]
"unique" was the wrong word---I really meant "standard, unified" way of updating temperatures.

Could `AndersenThermostat` pull the temperature from the `Context` or `Integrator` each time it needs to do an update?  

I'm just really worried that we make it very easy to do the *wrong* thing right now by having too many ways of setting the temperature.

> I'm also kind of uncomfortable about putting a setTemperature() method onto integrators that don't have anything to do with temperature, like VerletIntegrator.

If we want to centralize an authoritative temperature, it essentially either has to be at the `Integrator`, `System`, or `Context` level.

Putting the authoritative temperature in `Context` could work, but is inherently ephemeral since it can't be serialized.  Putting it in `System` could certainly work, though `Integrator` seems most logical.

**Comment by jchodera** [comment]
> And then what about DrudeLangevinIntegrator, which has two different temperatures?

Is that an official integrator?  I don't see it in [the API headers](https://github.com/pandegroup/openmm/tree/master/openmmapi/include/openmm).

The `DrudeLangevinIntegrator` is not really a two-temperature integrator.  It's isomorphic with a canonical system with a single true temperature in which the interactions with the Drude particles are scaled up by some factor related to `Tsystem / Tdrude`.  There's nothing magical about the second temperature---it's entirely fictitious and not a real temperature.

**Comment by peastman** [comment]
> Is that an official integrator? I don't see it in the API headers.

Yes.  It's a plugin, but it's one of the standard bundled ones.  It's as official as, say, AMOEBA.

> If we want to centralize an authoritative temperature, it essentially either has to be at the Integrator, System, or Context level.

If we're to store it in one of those places, Context sort of "feels" like the best place.  Anything that's expected to change over a simulation really belongs in Context.  I don't know what you mean about serialization.  Why would it be harder to serialize than all the other information stored in Contexts?  It would just be included in the State.

On the other hand, that would create various problems, both practical and conceptual.  The biggest practical problem is how to maintain backward compatibility.  This is a change to a very central feature, so I don't see  how to do this without breaking almost everyone's scripts.

On the conceptual side, this perhaps unifies the concept of temperature a little *too* much.  For example, is temperature really any more fundamental than pressure?  (Based on the title you chose for this issue, you don't seem to think so.)  If we centralize temperature, shouldn't we centralize pressure too?  Yet pressure clearly has to be represented in different ways depending on what sort of barostat you use.  It might be a single number, or three numbers, or perhaps there's also a surface tension in there.  Likewise, there need not always be a single temperature.  Dual Langevin integrators like for Drude particles are one example.  Colored noise integrators are another one.  Or perhaps someone wants to study heat flow, so they couple different parts of the system to different heat baths.  That isn't very common with biological macromolecules, but OpenMM also gets used for other things.  And of course, there's the very common case where there's no temperature at all, because you're doing a constant energy simulation.

Another possible approach to this problem is to define a common *internal* API for reporting temperature.  There wouldn't be one canonical place where temperature was always stored, but there would be a consistent mechanism so the barostat could discover it from wherever it was stored.

**Comment by jlmaccal** [comment]
I'm running into this issue now. What is the current correct way to change the temperature for an existing `MonteCarloBarostat`?

And just to clarify, `setTemperature` for a `LangevinIntegrator` does update the temperature in an existing context, correct?

**Comment by peastman** [comment]
What is the current correct way to change the temperature for an existing MonteCarloBarostat?

```Python
context.setParameter(MonteCarloBarostat.Temperature(), temperature)
```

> And just to clarify, setTemperature for a LangevinIntegrator does update the temperature in an existing context, correct?

Correct.

**Comment by jhprinz** [comment]
It's been a while I see. But this recently came to my attention. Mostly to access the current temperature for logging purposes. I am brainstorming a little to get the right picture here:

I realize there are two conceptual different ways to achieve a _thermostat_ or _barostat_: Either using a modified integrator (that might internally add forces) or just additional forces. I am mostly familiar with using forces and my intuition is that _in theory_ all integrator approaches could be reformulated by rewriting it into additional forces. Not saying that makes sense to do that. 

If that would be the case it would be great to have integrators, model forces and ensemble forces separated, right? And the ensemble forces could share common ensemble intensive properties like `temperature` and `pressure`. This is what @peastman means with: Context _feels_ the right place to put temperature. I agree.

Now, this is not possible and we have to deal with the 2 concepts and both integrators and/or forces may or may not be dependend on the ensemble temperature which we would compare to the laboratory conditions when a system is simulated. 

> Another possible approach to this problem is to define a common internal API for reporting temperature. There wouldn't be one canonical place where temperature was always stored, but there would be a consistent mechanism so the barostat could discover it from wherever it was stored.

This seems to be the right way then for now. Let me rephrase this idea a little: 

All objects (integrator/additional _thermo/barostating_ forces) that require or understand the concept of the laboratory temperature have a `setTemperature` and `getTemperature` function.
That would at least make it easy for a user to set the temperatures in these cases. You just know that `setTemperature` will affect the ensemble temperature. The context could then have a `setTemperature` function that delegates this to the integrator and forces/system just for convenience. Same for pressure.

The other point that makes sense for me is to have a second set of context forces which are treated the same as regular forces but they are not part of the system object. This is really just to put some forces under a separate name. This way you can avoid a change in pressure affect a system and so reuse the system object in several contexts. This will require a bigger change, but it will not break backwards-compatability.

> how we can make it easier for users to correctly set the temperature and pressure without allowing the potential for making simple mistakes that lead to incorrect simulations.

+1

**Comment by peastman** [comment]
> This way you can avoid a change in pressure affect a system and so reuse the system object in several contexts. This will require a bigger change, but it will not break backwards-compatability.

This part has already been addressed by changing the API for the barostats.  They now use a context parameter to store the current value of the pressure.  So you can change it without modifying the system just by writing

```Python
context.setParameter(MonteCarloBarostat.Pressure(), pressure)
```

**Comment by jchodera** [comment]
In light of the recent interest in this thread, I'd love to see standard get/set temperature methods. This would allow integrators to be much more modular, requiring fewer special cases to handle certain kinds of integrators.

For CustomIntegrator, setting the temperature could (1) populate a special variable called "temperature", or (2) allow the user to specify some update equations that get called if the temperature is changed via a set temperature() call. Option (2) might be optimal.

I don't think it's a significant concern that some integrators wouldn't do anything with the temperature information. After all, some custom integrators don't use constraints, but we still have a constraints tolerance you can set/get.

The biggest concern is that this might break some custom integrators in the wild, so it would be best if this was coordinated with a major version number update.

**Comment by Olllom** [comment]
Do you think that it would be a good idea to add a couple of high-level API properties to the simulation class to unify and simplify some of these settings for many use cases?

Something in that spirit:
```
    @property
    def positions(self):
        return self.context.getState(getPositions=True).getPositions()

    @positions.setter
    def positions(self, pos):
        self.context.setPositions(pos)

    @property
    def velocities(self):
        pass

    @velocities.setter
    def velocities(self, vel):
        pass

    @property
    def box(self):
        pass

    @box.setter
    def box(self, box_lengths):
        pass

    @property
    def temperature(self):
        pass

    @temperature.setter
    def temperature(self, temp):
        # set temperature for integrators, barostats
        pass

    @property
    def stepsize(self):
        pass

    @stepsize.setter
    def stepsize(self, step):
        pass

    @property
    def pressure(self):
        pass

    @pressure.setter
    def pressure(self, pres):
        pass

    def __setattr__(self):
        """Set context parameters as simulation.parameter_name = ..."""
        pass

    def __getattr__(self):
        """Get context parameters"""
        pass
```

**Comment by peastman** [comment]
I'm not sure.  On the one hand, it would make some code a little simpler.  On the other hand, it would encourage people to write their code in inefficient ways, since retrieving multiple pieces of information through a single call to `getState()` is faster than making multiple calls to it.

Some of the properties you suggested aren't clear cut to get or set.  For example, what exactly is the "temperature"?  You might have a LangevinIntegrator, in which we know how to get its temperature.  Also if your system includes an AndersenThermostat.  But if you have a DrudeLangevinIntegrator there are two different temperatures.  And if you're using a CustomIntegrator, we have no idea whether it's thermostated, or how to retrieve the temperature if it is.

**Comment by Olllom** [comment]
Sure. The scope of such an API would not cover all but the most common cases. 

Setting the temperature could look something like this:

```
    @temperature.setter
    def temperature(self, temp):
        is_temperature_set = 0
        BarostatClasses = [
            MonteCarloBarostat, 
            MonteCarloAnisotropicBarostat, 
            MonteCarloMembraneBarostat
        ]
        if hasattr(self.integrator, "setTemperature"):
            self.integrator.setTemperature(temp)
            is_temperature_set += 1
        for force in self.system.getForces():
            if isinstance(force, AndersenThermostat):
                self.context.setParameter(AndersenThermostat.Temperature(), temp)
                is_temperature_set += 1
            for Barostat in BarostatClasses:
                if isinstance(force, Barostat):
                    self.context.setParameter(Barostat.Temperature(), temp)
        if is_temperature_set == 0:
            raise Exception("Did not find a method to set the temperature.")
        elif is_temperature_set > 1:
            warnings.warn(
                "Temperature set in multiple places. "
                "Does the simulation have multiple thermostats?"
            )
```

This would support most thermostated custom integrators by duck-typing `setTemperature`.

The Drude temperature would of course not be covered but why should anybody bother changing it during a simulation anyway?
